File: cpp/ModUtil.cpp
//@verified http://codeforces.com/contest/559/submission/33876738
template <typename T>
class ModUtil
{
  public:
    ModUtil(int N, T mod) : N(N), mod(mod), fact(N + 1), inv_fact(N + 1) { calc(); }

    T get_inv(int x) { return modpow(x, mod - 2); }
    T get_fact(int n) { return fact[n]; }
    T get_inv_fact(int n) { return get_inv(fact[n]); }
    T get_P(int n, int k)
    {
        if (n < k)
            return 0;
        return fact[n] % mod * get_inv_fact(n - k) % mod;
    }
    T get_C(int n, int k)
    {
        if (n < k)
            return 0;
        return fact[n] * inv_fact[k] % mod * inv_fact[n - k] % mod;
    }
    T modpow(int a, int n)
    {
        if (n == 0)
            return 1;
        T res = modpow(a, n / 2);
        res *= res;
        res %= mod;
        if (n % 2)
            res *= a;
        return res % mod;
    }

  private:
    vector<T> fact;
    vector<T> inv_fact;
    int N;
    T mod;
    // O(N)
    void calc()
    {
        fact[0] = 1;
        for (int i = 1; i <= N; i++)
        {
            fact[i] = i * fact[i - 1] % mod;
        }
        inv_fact[0] = 1;
        for (int i = 1; i <= N; i++)
        {
            inv_fact[i] = modpow(fact[i], mod - 2);
        }
    }
};

File: rust/mod_util.rs
pub struct ModUtil {
    fact: Vec<usize>,
    inv_fact: Vec<usize>,
    module: usize,
}

#[allow(dead_code)]
impl ModUtil {
    pub fn new(size: usize, module: usize) -> ModUtil {
        let mut fact = vec![0; size + 1];
        let mut inv_fact = vec![0; size + 1];
        fact[0] = 1;
        for i in 1..size + 1 {
            fact[i] = i * fact[i - 1] % module;
        }
        inv_fact[0] = 1;
        for i in 1..size + 1 {
            inv_fact[i] = mod_pow(fact[i], module - 2, module);
        }
        ModUtil {
            fact: fact,
            inv_fact: inv_fact,
            module: module,
        }
    }

    pub fn permutation(&self, n: usize, k: usize) -> usize {
        assert!(n >= k);
        self.fact[n] % self.module * self.inv_fact(n - k) % self.module
    }
    pub fn combination(&self, n: usize, k: usize) -> usize {
        assert!(n >= k);
        self.fact[n] * self.inv_fact[k] % self.module * self.inv_fact[n - k] % self.module
    }

    pub fn fact(&self, x: usize) -> usize {
        self.fact[x]
    }
    pub fn inv(&self, x: usize) -> usize {
        mod_pow(x, self.module - 2, self.module)
    }
    pub fn inv_fact(&self, x: usize) -> usize {
        self.inv(self.fact[x])
    }
}

fn mod_pow(x: usize, n: usize, module: usize) -> usize {
    if n == 0 {
        return 1;
    }
    let mut res: usize = mod_pow(x, n / 2, module);
    res = (res % module * res % module) % module;
    if n % 2 == 1 {
        res *= x;
    }
    res % module
}